### PR TITLE
Refactor command-line interface

### PR DESCRIPTION
**Pull request by hackedy**
This PR makes the petr4.exe command-line interface more uniform and hopefully more understandable. I've created -output-LANG flags for each IR which accept a filename as argument and choose what kind of pretty printer to use based on the extension of the filename. So if you write `-output-p4light file.v` you get a Gallina version of the AST, but `-output-p4light file.p4` will print concrete P4 syntax.

```
% dune exec -- petr4 -help
Petr4: A reference implementation of the P4_16 language

  petr4 FILE.P4

=== flags ===

  [-I dir] ...               . Paths to search for files sourced with #include
                               directives.
  [-gen-loc]                 . Infer locators in P4light AST after typechecking.
  [-normalize]               . Simplify expressions.
  [-output-c file]           . Output C to the specified file.
  [-output-gcl file]         . Output GCL to the specified file.
  [-output-p4cub file]       . Output P4Cub to the specified file.
  [-output-p4flat file]      . Output P4flat to the specified file.
  [-output-p4light file]     . Output P4light to the specified file.
  [-output-p4surface file]   . Output P4surface AST to the specified file.
  [-unroll-parsers depth]    . Unroll parsers to given depth.
  [-v]                       . Be more verbose.
  [-build-info]              . print info about this build and exit
  [-version]                 . print the version of this build and exit
  [-help], -?                . print this help text and exit
```

### CODE REVIEW

**Comment by hackedy** [comment]
Right now it's just compiler passes, not evaluation, and I haven't tried to iron any bugs out yet.

The "p4flat" IR does not exist yet but I know what I want it to look like (sort of like Inline.v from the GCL compiler but with declarations).

TODOs
- [ ] add interpreter to CLI
- [ ] fix STF test driver
- [ ] add validator for compiler_cfg type
- [ ] add printing for Common.error messages

File: bin/main.ml [hunk 1]
```diff
@@ -17,116 +17,106 @@ open Core
...
-     +> flag "-pkt-str" (required string) ~doc: "<pkt_str> Add packet string"
-     +> flag "-port" (optional_with_default "0" string) ~doc: "<port_number> Specify ingress port"
-     +> flag "-T" (optional_with_default "v1" string) ~doc: "<target> Specify P4 target (v1, ebpf currently supported)"
-     +> anon ("p4file" %: string))
-    (fun verbose include_dir pkt_str port target p4file () ->
-       print_string (eval_file_string include_dir p4file verbose pkt_str (int_of_string port) target))
+  let close_file ochan =
+    match Core_unix.close_process_out ochan with
+    | Ok () -> ()
+    | Error _ -> failwith "Error in close_file"
```
Comment (zachary0kent) [inline bin/main.ml:42]:
```suggestion
  let close_file ochan =
    ochan 
    |> Core_unix.close_process_out 
    |> Result.ok_or_failwith "Error in close_file"
```

File: bin/main.ml [hunk 2]
```diff
@@ -17,116 +17,106 @@ open Core
...
-       ignore (compile_file include_dir p4file normalize export_file verbose gen_loc printp4cub printp4_file gcl))
+module UnixDriver = MakeDriver(UnixIO)
 
-let do_stf _ _ _ = failwith "do_stf unimplemented"
+let parse_ext_flag p =
+  match p with
+  | Some file ->
+     Pass.Run (Some (Pass.parse_output_exn file))
+  | None ->
+     Pass.Run None
```
Comment (zachary0kent) [inline bin/main.ml:53]:
```suggestion
let parse_ext_flag p = Pass.run (Option.map p ~f:Pass.parse_output_exn)
```

File: lib/pass.ml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    out_fmt: fmt }
+
+type 'a cfg =
+  | Skip
+  | Run of 'a
+
+let is_skip : 'a cfg -> bool =
+  function
+  | Skip -> true
+  | _ -> false
```
Comment (zachary0kent) [inline lib/pass.ml:20]: It seems like this could probably be dervied

File: bin/main.ml [hunk 4]
```diff
@@ -17,116 +17,106 @@ open Core
...
-        do_stf include_dir stf_file p4_file)
+let parse_backend unroll_parsers output_gcl output_clight =
+  match output_gcl, output_clight with
+  | Some gcl, None ->
+     begin match unroll_parsers with
+     | Some depth ->
+        let gcl_output = Pass.parse_output_exn gcl in
+        Pass.Run (Pass.GCLBackend {depth; gcl_output})
+     | None ->
+        failwith "please provide a depth with -unroll-parsers when compiling to GCL"
```
Comment (zachary0kent) [inline bin/main.ml:63]:
```suggestion
     let message = "please provide a depth with -unroll-parsers when compiling to GCL" in
     let depth = Option.value_exn unroll_parsers ~message in
     let gcl_output = Pass.parse_output_exn gcl in
     Pass.Run (Pass.GCLBackend {depth; gcl_output})
```

File: bin/main.ml (hunk 1 above)
Comment (hackedy) [reply bin/main.ml:42]: Sweet

File: lib/pass.ml (hunk 3 above)
Comment (hackedy) [reply lib/pass.ml:20]: what do you mean?

**Comment by hackedy** [comment]
I think this is fully baked now.

**Pull request merged**